Add an endpoint for getting pay for order orders

hsingyuc opened the pull request:
<!-- Start by describing the changes made in this Pull Request and the reason for such changes. -->
Add an endpoint for getting pay-for-order orders so we can use that to hydrate an order block in the checkout.



<!-- Reference any related issues or PRs here -->

Fixes #8922 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Other Checks

- [x] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots
<img width="499" alt="Screen Shot 2023-06-28 at 6 39 15 PM" src="https://github.com/woocommerce/woocommerce-blocks/assets/56378160/09fae239-e73d-49d0-94bf-841e51c85bb7">

### Testing
1. Install and activate a plugin with pay-for-order flow. ex WooCommerce-Bookings
3. Create a product that needs confirmation
1. Purchase the product with pay-for-order flow
1. Go to WooCommerce -> Orders
1. Add some order fees for tax, shipping, etc
4. Click `Customer payment page` to go to the pay-for-order page
1. Copy the order_id and the key from the URL
1. Use Postman to test this endpoint `http://merchant.local/wp-json/wc/store/order/YOUR_ORDER_NUMBER?key=YOUR_KEY&billing_email=YOUR_BILLING_EMAIL`
2. Check that the order items, totals, ID, etc, are there

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests



### Changelog

> Add an endpoint for getting pay-for-order orders

hsingyuc committed:
Register order route

hsingyuc committed:
Get order it

hsingyuc committed:
Add order schema

hsingyuc committed:
Check authorization for getting the order

hsingyuc committed:
Add order data to the response

hsingyuc committed:
Add order schema for the endpoint

hsingyuc commented on src/StoreApi/Schemas/V1/OrderSchema.php:
```diff
@@ -0,0 +1,161 @@
+		return [
+			'id'         => $order->get_id(),
+			'line_items' => $this->get_item_data( $order ),
+			'totals'     => $order->get_order_item_totals(),
```
I initially manually added all the totals here.  The `get_order_item_totals` method includes HTML and currencies so we'll need to dangerously set and sanitize in React, but it uses more filters from WordPress which could be useful.

I'm not sure which one is better.

senadir replied:
regardless of how we continue with this PR, endpoint response should be consisent with the rest of requsts, so items instead of line items and it should be formated via CartItemSchema, same for totals...

hsingyuc replied:
Thanks for the comment, @senadir.  I'm not sure that I understand your suggestion here.  

>endpoint response should be consisent with the rest of requsts, so items instead of line items and it should be formated via CartItemSchema, same for totals...

It seems like the solution we decided not to go with.  That solution was to match order data to the cart schema so we could use the existing cart block.  Did I understand your suggestion correctly?

After this PR adds the pay-for-order order endpoint, [we will create a `CheckoutOrderPayForOrderBlock` to hydrate the order using this endpoint.](https://href.li/?https://github.com/Automattic/woocommerce-blocks/blob/b1e07428a8858f7c76c41bc15c87d23f39cb01d5/src/BlockTypes/Checkout.php#L493)  That's why we don't convert the order schema to the cart item schema here.

hsingyuc committed:
Move validation check to order controller

hsingyuc committed:
Update the error codes

hsingyuc committed:
Add order item schema

mikejolley commented on src/StoreApi/Routes/V1/Order.php:
```diff
@@ -0,0 +1,107 @@
+	 * @param \WP_REST_Request $request Request object.
+	 * @return boolean|WP_Error
+	 */
+	public function is_authorized( \WP_REST_Request $request ) {
```
Let's include an additional check for user ID if the order is associated with a user.

hsingyuc committed:
Check if the order is associated with current user

hsingyuc committed:
Update order schema to match cart schema

hsingyuc committed:
Update order item schema to match cart item schema

hsingyuc committed:
Add product item trait

hsingyuc committed:
Update sold individually property

hsingyuc committed:
Allow guests to pay for order

hsingyuc committed:
Update wording for logged out customers

hsingyuc commented:
@mikejolley I made the changes for this PR, and for #9401 to be able to use the `validate_order_key`, we need this PR to be merged first.   Also, I changed the base branch to `add/pay-for-order-flow`.  Can you review when able to unblock other work?  Thank you!

mikejolley commented on src/StoreApi/Routes/V1/Order.php:
```diff
@@ -0,0 +1,109 @@
+		$order     = wc_get_order( $order_id );
+
+		try {
+			$this->order_controller->validate_order_key( $order_id, $order_key );
```
I made a comment in `validate_order_key` that it needs to be split between order key validation, and payment validation. We are still missing one check which should be added here.

If the order ID has a customer ID associated with it, **check that the current user ID is a match**.  If not, reject the request. Since they would have had the correct key due to the above check, we could probably be lax with the error message and say "Invalid order. This order belongs to another customer." or similar. What do you think?

If the order does not have a customer ID associated with it, skip this check.

mikejolley replied:
Looking at source, I can see now that `pay_for_order` cap does this. It's naming isn't very clear—maybe we just move the check inline and avoid  `pay_for_order`—I don't think `pay_for_order` is appropriate to determine if an order can be returned here.

hsingyuc replied:
I moved `current_user_can` inline, and updated the function to `get_failed_order_stock_error` and added this to the `errors` response so we can use it in the frontend later. [239620c](https://github.com/woocommerce/woocommerce-blocks/pull/8960/commits/239620c0184e92693e70db9782d8969e067d7b93)

mikejolley replied:
Thanks—I'm pushing a followup to remove the duplicate `pay_for_order` check and some other issues I noticed when testing.

mikejolley replied:
085933c6f36c165a09d0ba3722a22fb3928964c1

hsingyuc replied:
Thank you, Mike!!

mikejolley commented on src/StoreApi/Schemas/V1/OrderItemSchema.php:
```diff
@@ -0,0 +1,381 @@
+	 *
+	 * @return array
+	 */
+	public function get_properties() {
```
To reduce the likelihood of mismatching schema, do you think we should create an abstract item schema and have `OrderItemSchema`/`CartItemSchema` consume it? We'd have to remove any references to "cart" in the property descriptions but it should be possible to make those descriptions more generic or applicable to both.

That way we only need to maintain a single list of schema.

hsingyuc replied:
Good idea. Create an abstract class item schema here  b36e297.

mikejolley commented on src/StoreApi/Schemas/V1/OrderItemSchema.php:
```diff
@@ -0,0 +1,381 @@
+	 */
+	public function get_totals( $item ) {
+		return [
+			'line_subtotal'     => $item->get_subtotal(),
```
Are these missing a call to `prepare_money_response`?

hsingyuc replied:
Thank you for pointing that out!  b36e297

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+		$order = $this->get_order( $order_id );
+
+		if ( ! $order || $order->get_id() !== $order_id || ! hash_equals( $order->get_order_key(), $order_key ) ) {
+			throw new RouteException( 'woocommerce_rest_invalid_order', __( 'Sorry, this order is invalid and cannot be paid for.', 'woo-gutenberg-products-block' ), 401 );
```
We should make these error messages more generic since this endpoint may not be used just for getting details for a payment. Also, we should avoid revealing if an order ID is valid or not to a user without authentication. Any user facing messages should be handled by the client.

mikejolley replied:
Addressed in f7e70aa305ed20b2abcee1889d74fcc51d1bf7ab

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+	 * @param integer $order_id Order ID.
+	 * @param string  $order_key Order key.
+	 */
+	public function validate_order_key( $order_id, $order_key ) {
```
This method seems to be serving 2 functions;

1. To check if the order key is valid
2. To see if an order is payable

I'm not sure it should handle both—we should split it and only validate payment if the user is trying to pay somehow? Or perhaps include that in the response. I'm not 100% sure yet.

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+		$order = $this->get_order( $order_id );
+
+		if ( ! $order || $order->get_id() !== $order_id || ! hash_equals( $order->get_order_key(), $order_key ) ) {
+			throw new RouteException( 'woocommerce_rest_invalid_order', __( 'Sorry, this order is invalid and cannot be paid for.', 'woo-gutenberg-products-block' ), 401 );
```
```suggestion
			throw new RouteException( 'woocommerce_rest_invalid_order', __( 'Invalid order ID or key.', 'woo-gutenberg-products-block' ), 401 );
```

mikejolley replied:
Addressed in f7e70aa305ed20b2abcee1889d74fcc51d1bf7ab

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+		}
+
+		// Logged out customer does not have permission to pay for this order.
+		if ( ! current_user_can( 'pay_for_order', $order_id ) && ! is_user_logged_in() ) {
```
We need to find a better place for this logic. I'm not sure whether we need to somehow declare that we're getting an order for payment (intent), or if we should return something in the response stating if payment is possible or not for the current user context, or if we validate when you attempt payment only.

The main reason is, say we are getting an order to view the details on the thanks page. We don't care about payment in that context.

What are your thoughts on this?

I will comment on the rules below anyway.

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+		}
+
+		// Does not need payment.
+		if ( ! $order->needs_payment() ) {
```
Sounds like this would be something useful to return in the response?

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+		// Does not need payment.
+		if ( ! $order->needs_payment() ) {
+			/* translators: %s: order status */
+			throw new RouteException( 'woocommerce_rest_no_payment_needed', sprintf( __( 'This order&rsquo;s status is &ldquo;%s&rdquo;&mdash;it cannot be paid for. Please contact us if you need assistance.', 'woo-gutenberg-products-block' ), wc_get_order_status_name( $order->get_status() ) ), 403 );
```
This error message is fine, if the context is making a payment. If we're just getting order details the check is not appropriate.

mikejolley commented on src/StoreApi/Utilities/OrderController.php:
```diff
@@ -461,6 +473,104 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
+			throw new RouteException( 'woocommerce_rest_no_payment_needed', sprintf( __( 'This order&rsquo;s status is &ldquo;%s&rdquo;&mdash;it cannot be paid for. Please contact us if you need assistance.', 'woo-gutenberg-products-block' ), wc_get_order_status_name( $order->get_status() ) ), 403 );
+		}
+
+		// Ensure order items are still stocked if paying for a failed order. Pending orders do not need this check because stock is held.
```
O I've never seen this logic before. I see its included in the shortcode checkout. I think that logic is flawed—if we allow failed orders to be reattempted but don't hold the stock that feels like a bug that should be fixed in `get_query_for_reserved_stock` in core.

I think it might be good to fix this is core so we don't need to adopt this complex logic in the store API. So log it upstream and/or patch `get_query_for_reserved_stock` to keep stock reserved for pending, failed, and `wc-checkout-draft` orders.

hsingyuc replied:
I think failed orders shouldn't hold stock.  If shoppers don't come back to finish the checkout, then we'll hold the stock forever, right?

I added this check to `errors` so it doesn't block the endpoint, but do agree that we should handle this logic upstream.  I can remove this for now if you think it's better.

mikejolley replied:
> I think failed orders shouldn't hold stock

If the customer can retry within a certain time limit that would be a better experience for them that being locked out. Stock does get freed up after a certain time limit. Pending orders have this same treatment—the order could be abandoned at that point also.

hsingyuc replied:
Sounds good. Let's handle this upstream.  What do you think?

mikejolley reviewed (changes requested):
I've added some feedback—the main thing we need to consider is that this endpoint is to get order details, not necessarily to pay for an order. We need the payment logic to either run only when you're paying, or a way to tell the endpoint that we're getting the order in payment context rather than view context. I'm not sure of the best way forward with this just yet. Let me know your thoughts.

mikejolley reviewed (changes requested):
I've added some feedback—the main thing we need to consider is that this endpoint is to get order details, not necessarily to pay for an order. We need the payment logic to either run only when you're paying, or a way to tell the endpoint that we're getting the order in payment context rather than view context. I'm not sure of the best way forward with this just yet. Let me know your thoughts.

hsingyuc committed:
Allow getting all orders from the endpoint

mikejolley committed:
Add inline explanation for pay_for_order capability

mikejolley committed:
Remove unused $user_id and $order variables

mikejolley committed:
Remove duplicate pay_for_order capability check from validate_order_key

mikejolley committed:
Update exception wording when missing order id or key, or user mismatch

mikejolley committed:
Ensure $order_key is not null to avoid fatal error when left blank

mikejolley committed:
Resolve linting errors in order route class

mikejolley committed:
Adjust order ID description

mikejolley commented:
@wavvves if you look at this while I'm AFK, we need to review the schema. Ideally we can keep cart/order/checkout schema similar (in particular item and totals types), but some cart properties may not apply well to orders. For example, cart has `has_calculated_shipping`, but isn't needed for orders afaik. cc @hsingyuc

hsingyuc committed:
Create an abstract item schema

hsingyuc commented:
@mikejolley @wavvves I've made almost all of the changes except [this one](https://github.com/woocommerce/woocommerce-blocks/pull/8960#discussion_r1214918661).  Could you take a look when you have time?  Thank you!

wavvves commented:
I'm sorry for being late to this conversation, read the whole (!) discussion here and on Slack, but apologies in advance in case I miss something and my suggestions are redundant or already discussed.

> Add an endpoint for getting pay-for-order orders so we can use that to hydrate an order block in the checkout.

This sentence got stuck in my head from the start. First, do we need an endpoint specifically for "getting pay-for-order orders"? Although not available immediately, this fits into a broader context: account management.
The purpose is to hydrate an order block, so to make this granular enough, we should consider that this is a simple use case for the orders endpoint.

It is expectable that we need an order list and order detail in the future. And these make part of a set of endpoints that should live inside auth, although not cookie-based. Since that is not available yet, there is no problem with doing an id/key combo to secure these. Order listing, on the other hand, must have auth. To sum it up, discussing the future of order data endpoints and namespacing should go hand in hand with response schemas. Having this available first is not a problem, as long as it fits future needs for my account.

hsingyuc commented:
Thank you for taking the time to read the discussion, @wavvves!

>This sentence got stuck in my head from the start. First, do we need an endpoint specifically for "getting pay-for-order orders"?

I think after the changes above, the endpoint is not specific to pay for orders anymore and the response is a generic order response.

The only piece that might not work for an order list is the "key" but I think we can tweak this later if we want to allow authorized users to access without a key.  For now, I think leaving this key in keeps things more secure.

wavvves commented:
> there is no problem with doing an id/key combo to secure these.

Disregard what I said here, actually this might not be true in the near future. I updated the slack thread with a link that explains why.

p1686394628493249/1680686053.947699-slack-C8X6Q7XQU

mikejolley commented:
Thanks for the changes @hsingyuc. The remaining blockers are:

- [Decision on stock handling or making changes upstream](https://github.com/woocommerce/woocommerce-blocks/pull/8960#discussion_r1214918661)
- The p2 discussion about requiring the email address for guest orders
- Agreeing on order schema. @wavvves We could use your input on this too.

Just to get a high level view of the current proposed cart and order schema, I plugged it into Chat CPT to generate the following.

| Item Name                 | Type     | Order Description                                                                                                         | Cart Description                                                                                                       | Notes                                                                                                                            |
|---------------------------|----------|--------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|
| id                        | integer  | The order ID.                                                                                                            | N/A                                                                                                                    |                                                                                                                                  |
| items                     | array    | Line items data.                                                                                                         | List of cart items.                                                                                                    |                                                                                                                                  |
| totals                    | object   | Order totals.                                                                                                            | Cart total amounts provided using the smallest unit of the currency.                                                   |                                                                                                                                  |
| coupons                   | array    | List of applied cart coupons.                                                                                            | List of applied cart coupons.                                                                                          |                                                                                                                                  |
| shipping_rates            | array    | List of available shipping rates for the cart.                                                                           | List of available shipping rates for the cart.                                                                         | For orders, this is an empty array.                                                                                               |
| shipping_address          | object   | Current set shipping address for the customer.                                                                           | Current set shipping address for the customer.                                                                         |                                                                                                                                  |
| billing_address           | object   | Current set billing address for the customer.                                                                            | Current set billing address for the customer.                                                                          |                                                                                                                                  |
| items_count               | integer  | Number of items in the cart.                                                                                             | Number of items in the cart.                                                                                           |                                                                                                                                  |
| items_weight              | number   | Total weight (in grams) of all products in the cart.                                                                     | Total weight (in grams) of all products in the cart.                                                                   | For orders, this is 0.                                                                                                            |
| cross_sells               | array    | List of cross-sells items related to cart items.                                                                         | List of cross-sells items related to cart items.                                                                       | For orders, this is an empty array.                                                                                               |
| needs_payment             | boolean  | True if the cart needs payment. False for carts with only free products and no shipping costs.                          | True if the cart needs payment. False for carts with only free products and no shipping costs.                        |                                                                                                                                  |
| needs_shipping            | boolean  | True if the cart needs shipping. False for carts with only digital goods or stores with no shipping methods set-up.       | True if the cart needs shipping. False for carts with only digital goods or stores with no shipping methods set-up.     |                                                                                                                                  |
| has_calculated_shipping   | boolean  | True if the cart meets the criteria for showing shipping costs, and rates have been calculated and included in the totals.| True if the cart meets the criteria for showing shipping costs, and rates have been calculated and included in the totals.| For orders, this is null.                                                                                                        |
| fees                      | array    | List of cart fees.                                                                                                      | List of cart fees.                                                                                                    | For orders, this is an empty array.                                                                                               |
| errors                    | array    | List of cart item errors, for example, items in the cart which are out of stock.                                         | List of cart item errors, for example, items in the cart which are out of stock.                                       |                                                                                                                                  |
| payment_requirements      | array    | List of required payment gateway features to process the order.                                                          | List of required payment gateway features to process the order.                                                        |                                                                                                                                  |

There are some properties that have no values in the order schema—we could either populate them from the order, or remove them from the schema if not needed. The data needs to be consistent between order and cart, but we don't necessarily have to include all properties. 

These are the (currently unpopulated) properties in orders to consider:

- `shipping_rates`
- `items_weight` - Only used for shipping rates in cart?
- `cross_sells` - Not needed for orders
- `has_calculated_shipping` - Only used for shipping rates in cart?
- `fees`

`items_count` is used for shipping I believe? So if we removed the shipping rate properties above this could be removed also.

In the context of orders, fees, discounts, and shipping are stored as `line items`. Do we need to include these in the response? You can refer to the main WC Rest API to see how its handled there: https://woocommerce.github.io/woocommerce-rest-api-docs/#order-properties

@hsingyuc Would WooPay need fees + shipping + discounts etc in the shopper facing render of order data, or is that limited to only cart items and totals? I'm happy to leave these properties out if it leads to a simpler API. The totals may be enough.

hsingyuc committed:
Remove unused properties

hsingyuc commented:
> Decision on stock handling or making changes upstream

Answered [here](https://github.com/woocommerce/woocommerce-blocks/pull/8960#discussion_r1237985430)

> The p2 discussion about requiring the email address for guest orders

We are waiting on this PR (https://github.com/Automattic/woocommerce/ pull/273/) to get merged.


> There are some properties that have no values in the order schema—we could either populate them from the order, or remove them from the schema if not needed. The data needs to be consistent between order and cart, but we don't necessarily have to include all properties.

I recall we want to allow shoppers to be able to change their shipping address, if so, we will need below:
`shipping_rates`
`items_count` is used for shipping I believe? So if we removed the shipping rate properties above this could be removed also.
`items_weight`- Only used for shipping rates in cart?

Removed:
`cross_sells` - Not needed for orders
`has_calculated_shipping` - Only used for shipping rates in cart?
`fees` 

> In the context of orders, fees, discounts, and shipping are stored as line items. Do we need to include these in the response? You can refer to the main WC Rest API to see how its handled there: https://woocommerce.github.io/woocommerce-rest-api-docs/#order-properties

If we are allowing shoppers to change their addresses, then we need those in the response, correct?

> @hsingyuc Would WooPay need fees + shipping + discounts etc in the shopper facing render of order data, or is that limited to only cart items and totals? I'm happy to leave these properties out if it leads to a simpler API. The totals may be enough.

Are you asking if WooPay only needs total shipping, total fees, total discounts etc?  If we are not allowing shoppers to update their addresses, then yes, that's correct.

mikejolley commented:
> I recall we want to allow shoppers to be able to change their shipping address, if so, we will need below

I'd remove them to get this PR moving and re-implement them in schema when updating addresses is added/supported. I assume in a follow up if we're doing it.

hsingyuc committed:
Remove unused properties

hsingyuc commented:
>I'd remove them to get this PR moving and re-implement them in schema when updating addresses is added/supported. I assume in a follow up if we're doing it.

Sounds good. Removed here [6ce981f](https://github.com/woocommerce/woocommerce-blocks/pull/8960/commits/6ce981fba2447c211c39d7b015af9ad86f2c513b).

hsingyuc committed:
Add billing email validation

hsingyuc commented:
@mikejolley & @wavvves  The billing email address for guest orders PR is not merged yet but seems confirmed so I'm adding the validation in 
f048347

mikejolley commented:
Thanks @hsingyuc I'll get this one last review. Do you think we should try to wrap this new endpoint within a `if ( $this->package->is_experimental_build() ) {` call so it can only be used in the dev build? That way we won't ship it to users via feature plugin or core until we're 100% finished. Thoughts?

mikejolley commented:
@hsingyuc  I pushed a fix to handle coupons for orders and fees. To test, I simply applied a fixed cart coupon to the cart and used https://wordpress.org/plugins/woo-custom-fee/ to add a test fee. This should work for both cart and order now.

Do we need to add other order properties like `status`?

hsingyuc committed:
Allow to use the order endpoint in dev build only

hsingyuc commented:
> Do you think we should try to wrap this new endpoint within a if ( $this->package->is_experimental_build() ) { call so it can only be used in the dev build? That way we won't ship it to users via feature plugin or core until we're 100% finished. Thoughts?

That sounds safer 👍.  Do we still want to merge this into the feature branch or directly to `trunk` since it's behind this check now?

Thank you for providing the experimental condition.  Here's the change be42dda. 

Question about using `is_experimental_build`: when I test it, it doesn't seem to overwrite the `woocommerce_blocks_phase` to `1` or `2`.  Are we suppose to use `is_development_environment` instead?  Or should this only happen on `npm run build:deploy`?

hsingyuc committed:
Add order status property

hsingyuc commented:
> Do we need to add other order properties like status?

I was thinking we can add it in when needed but here's the change [86e3db5](https://github.com/woocommerce/woocommerce-blocks/pull/8960/commits/86e3db568cd144716fd1c9f0a1988ff2ce9ac7cd).

hsingyuc commented:
> @hsingyuc I pushed a fix to handle coupons for orders and fees. To test, I simply applied a fixed cart coupon to the cart and used https://wordpress.org/plugins/woo-custom-fee/ to add a test fee. This should work for both cart and order now.

@mikejolley Could you push your fix again?  I didn't see your fix and I force pushed my changes.  Thank you!

mikejolley commented:
> @mikejolley Could you push your fix again? I didn't see your fix and I force pushed my changes. Thank you!

Done!

> Question about using is_experimental_build: when I test it, it doesn't seem to overwrite the woocommerce_blocks_phase to 1 or 2. Are we suppose to use is_development_environment instead? Or should this only happen on npm run build:deploy?

I think it happens on deploy and building the release zip. I tested by manually setting `woocommerce_blocks_phase` to 1 in blocks.ini.

This looks good to me now. With the flag in place did could go into trunk I think. Pinging @woocommerce/rubik for an additional review.

hsingyuc commented:
@mikejolley On second thought, should we [also get the billing email from the session](273-gh-Automattic/woocommerce/files) so we don't need to pass the billing param in the update customer and checkout order endpoints?

mikejolley commented:
> @mikejolley On second thought, should we [also get the billing email from the session](273-gh-Automattic/woocommerce/files) so we don't need to pass the billing param in the update customer and checkout order endpoints?

That link didn't work. Could you elaborate please?

hsingyuc commented:
Sorry about that, I should have checked the link.  Tamper monkey updated incorrectly.  In the PR that implements the billing email check 273-gh-Automattic/woocommerce, they are getting the customer email from the session.  I think we can do the same.

```
if ( is_a( $session, WC_Session::class ) ) {
	$customer      = $session->get( 'customer' );
	$session_email = is_array( $customer ) && isset( $customer['email'] ) ? $customer['email'] : '';
}

$session_email_match  = $session_email === $order->get_billing_email();
```

mikejolley commented:
@hsingyuc Yeah that would work for us too 👍🏻 Then the param can be marked optional.

hsingyuc commented:
@mikejolley  I've been checking how WC sets and uses the session, and I'm not sure if it works with our use case.  It doesn't seem we are using the `$session_email` while checking `$session_email_match  = $session_email === $order->get_billing_email();`.  If you have more information please let me know, otherwise, I think we can keep the URL API approach.  Thank you!

hsingyuc committed:
Fix coupon and fee handling

mikejolley commented:
> I've been checking how WC sets and uses the session, and I'm not sure if it works with our use case. It doesn't seem we are using the $session_email while checking $session_email_match  = $session_email === $order->get_billing_email();. If you have more information please let me know, otherwise, I think we can keep the URL API approach.

Would it work to check to see if current user has a session and if so, grab their billing email address from session if it exists, then compare against billing email of order? If someone went through the flow on the store itself they'd have a session right? I'm think of ways we can make this endpoint work for logged out users using the website, not just API calls. Am I overthinking?

wavvves commented:
> We are waiting on this PR (https://github.com/Automattic/woocommerce/pull/273/) to get merged.

@hsingyuc just to be sure, we always provide billing email + key, and the way we protect not-guest orders from being accessed with that combo is by using 'pay_for_order' capability for the logged user, which should let it slide for orders made by a guest correct?

If so I think this is fine, and will be compatible when we want to extend this to be authed by JWT header tokens 👍🏼 

This PR looks good to me, needs tests to prevent it from breaking in the future, though. I'm fine merging it without temporarily if behind experimental.

@mikejolley is there any specific thing you are not still sure about and need input from me? Lights look 🟢

mikejolley commented:
@wavvves If you've looked over the code you should approve it too, focus on security aspects.

It is indeed featured flagged so I will pre-approve also.

hsingyuc commented:
@mikejolley & @wavvves  Thank you for reviewing and approving the PR!

> @hsingyuc just to be sure, we always provide billing email + key, and the way we protect not-guest orders from being accessed with that combo is by using 'pay_for_order' capability for the logged user, which should let it slide for orders made by a guest correct?

@wavvves We don't always provide billing email + key. We only need them when it's guest users.  And you are right about us using 'pay_for_order' capability for both types of users.

hsingyuc commented:
@mikejolley Just to be clear since we created a feature flag branch but also put these behind the dev build flag, would you like me to merge this PR to `trunk` or the `feature branch`?

ralucaStan commented:
> @mikejolley Just to be clear since we created a feature flag branch but also put these behind the dev build flag, would you like me to merge this PR to `trunk` or the `feature branch`?

@hsingyuc Considering this is behind a dev build flag we could merge this in trunk. 

Would you mind mentioning this on this page here by adding a new entry? https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md#features-flags-and-experimental-interfaces-

hsingyuc committed:
Update documentation for feature flags

hsingyuc committed:
Update typo in total fees tax

hsingyuc committed:
Update typo in tax lines

hsingyuc committed:
Add missing payment methods to cart response
